weaver/wps_restapi/jobs/utils.py

```diff
@@ -0,0 +1,699 @@
+                except Exception:
+                    err_code = None
+                if err_code:
+                    break
+            if not err_code:  # default
+                err_code = OWSNoApplicableCode.code
+                err_info = "unknown"
+            # /req/core/job-results-failed
+            raise HTTPBadRequest(json={
+                "title": "JobResultsFailed",
+                "type": err_code,
+                "detail": "Job results not available because execution failed.",
+                "status": HTTPBadRequest.code,
+                "cause": err_info,
+                "links": links
+            })
+
+        # /req/core/job-results-exception/results-not-ready
+        raise HTTPNotFound(json={
+            "title": "JobResultsNotReady",
+            "type": "http://www.opengis.net/def/exceptions/ogcapi-processes-1/1.0/result-not-ready",
+            "detail": "Job is not ready to obtain results.",
+            "status": HTTPNotFound.code,
+            "cause": {"status": job.status},
+            "links": links
+        })
+
+
+def raise_job_dismissed(job, container=None):
+    # type: (Job, Optional[AnySettingsContainer]) -> None
+    """
+    Raise the appropriate messages for dismissed :term:`Job` status.
+    """
+    if job.status == Status.DISMISSED:
+        # provide the job status links since it is still available for reference
+        settings = get_settings(container)
+        job_links = job.links(settings)
+        job_links = [link for link in job_links if link["rel"] in ["status", "alternate", "collection", "up"]]
+        raise JobGone(
+            json={
+                "title": "JobDismissed",
+                "type": "JobDismissed",
+                "status": JobGone.code,
+                "detail": "Job was dismissed and artifacts have been removed.",
+                "cause": {"status": job.status},
+                "value": str(job.id),
+                "links": job_links
+            }
+        )
+
+
+def dismiss_job_task(job, container):
+    # type: (Job, AnySettingsContainer) -> Job
+    """
+    Cancels any pending or running :mod:`Celery` task and removes completed job artifacts.
+
+    .. note::
+        The :term:`Job` object itself is not deleted, only its artifacts.
+        Therefore, its inputs, outputs, logs, exceptions, etc. are still available in the database,
+        but corresponding files that would be exposed by ``weaver.wps_output`` configurations are removed.
+
+    :param job: Job to cancel or cleanup.
+    :param container: Application settings.
+    :return: Updated and dismissed job.
+    """
+    raise_job_dismissed(job, container)
+    if job.status in JOB_STATUS_CATEGORIES[StatusCategory.RUNNING]:
+        # signal to stop celery task. Up to it to terminate remote if any.
+        LOGGER.debug("Job [%s] dismiss operation: Canceling task [%s]", job.id, job.task_id)
+        celery_app.control.revoke(job.task_id, terminate=True)
```

Comment:
Given that the revoke command is asynchronous, could the cleanup following it (removing logs, xml, dir) be done prematurily while the process is still spitting logs waiting to be killed?

Reply:
That is a possibility. I haven't explored this much.
If we encounter this, we could think of a way to handle it. Probably need to run another task with a delay to do a cleanup with below checks.

Reply:
In fact, the succeeded jobs also leave their stuff,  so a mechanism to clean that should already be in place anyway. For the typical birds dumping the result in the output directory, there was a cron job for that. Does it apply to Weaver also?

Reply:
The results that were generated by pywps and cwltool should be all cleaned up following execution by weaver after it moved them to wps-outputs. There should be 3 items `{jobId}.log`, `{jobId}.xml` (WPS status), `{jobId}/` (output results). 

The dismiss cleanup applied here is more specific for OGC Dismiss operation that requests explicitly a cleanup of those outputs, given a user downloaded them. 

If the cron job runs inside `wps-outputs/` for each *birds*, then it should pick up weaver as well, but there is not a cron specifically within weaver.